File: unowned-record-overwritten/node/check-record-ownership.js
/**
 * Detect a DNS record a reconciler is about to write without proof of
 * ownership, and repair it via Cloudflare only when the ownership marker
 * matches. Safe to run on a schedule. Stays in dry run until DRY_RUN=false.
 *
 * Guide: https://www.allanninal.dev/dns/unowned-record-overwritten/
 */
import { pathToFileURL } from "node:url";

const DNS_DOMAIN = process.env.DNS_DOMAIN || "example.com";
const CLOUDFLARE_API_TOKEN = process.env.CLOUDFLARE_API_TOKEN || "";
const CLOUDFLARE_ZONE_ID = process.env.CLOUDFLARE_ZONE_ID || "";
const DNS_OWNER_ID = process.env.DNS_OWNER_ID || "team-a";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

/**
 * intended: { name: string, type: string, content: string, owner: string }
 * live: { name: string, type: string, content: string, comment: string|null } or null if absent
 * ownerId: this reconciler's own owner tag, e.g. "team-a"
 * Returns one of: "create", "update", "skip_conflict", "noop"
 * Pure decision logic, no I/O: given the intended record, the current live record
 * (or null), and this instance's owner id, decide whether it is safe to write.
 */
export function decideAction(intended, live, ownerId) {
  if (live === null || live === undefined) {
    return "create";
  }

  const liveComment = live.comment || "";
  const liveOwner = liveComment.includes("managed-by:")
    ? liveComment.split("managed-by:").pop().trim()
    : null;

  if (liveOwner === null) {
    return "skip_conflict";
  }
  if (liveOwner !== ownerId) {
    return "skip_conflict";
  }
  if (live.content === intended.content) {
    return "noop";
  }
  return "update";
}

export async function run() {
  const zoneId = CLOUDFLARE_ZONE_ID;
  const apiToken = CLOUDFLARE_API_TOKEN;
  const ownerId = DNS_OWNER_ID;
  const dryRun = DRY_RUN;
  const domain = DNS_DOMAIN;

  // In a real run, "desired" would come from Terraform state or a
  // Kubernetes Ingress spec. Here it is one example record derived from
  // DNS_DOMAIN, so the script has something concrete to reconcile.
  const desired = [
    { name: `app.${domain}`, type: "A", content: "203.0.113.10", owner: ownerId },
  ];

  const headers = { Authorization: `Bearer ${apiToken}`, "Content-Type": "application/json" };
  const base = `https://api.cloudflare.com/client/v4/zones/${zoneId}/dns_records`;

  for (const intended of desired) {
    const listUrl = `${base}?name=${encodeURIComponent(intended.name)}&type=${intended.type}`;
    const listRes = await fetch(listUrl, { headers });
    if (!listRes.ok) throw new Error(`Cloudflare list failed: ${listRes.status}`);
    const listBody = await listRes.json();
    const results = listBody.result || [];
    const live = results.length > 0 ? results[0] : null;

    const action = decideAction(intended, live, ownerId);
    console.log(`Record ${intended.name} (${intended.type}): action=${action}`);

    if (action === "noop") {
      console.log("Already correct and owned. Nothing to do.");
      continue;
    }

    if (action === "skip_conflict") {
      console.warn(
        `Skipping ${intended.name}: live record has no matching owner marker for '${ownerId}'. ` +
        `Refusing to overwrite a record this reconciler does not own.`
      );
      continue;
    }

    const comment = `managed-by:${ownerId}`;
    if (action === "create") {
      console.log(`${dryRun ? "Would" : "Will"} create ${intended.name} -> ${intended.content}`);
      if (!dryRun) {
        const res = await fetch(base, {
          method: "POST",
          headers,
          body: JSON.stringify({ type: intended.type, name: intended.name, content: intended.content, ttl: 300, comment }),
        });
        if (!res.ok) throw new Error(`Cloudflare create failed: ${res.status}`);
      }
    } else if (action === "update") {
      console.log(`${dryRun ? "Would" : "Will"} update ${intended.name} -> ${intended.content}`);
      if (!dryRun) {
        const res = await fetch(`${base}/${live.id}`, {
          method: "PATCH",
          headers,
          body: JSON.stringify({ content: intended.content, comment }),
        });
        if (!res.ok) throw new Error(`Cloudflare update failed: ${res.status}`);
      }
    }
  }

  console.log("Done.");
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((e) => { console.error(e); process.exit(1); });
}

File: unowned-record-overwritten/python/check_record_ownership.py
"""Detect a DNS record a reconciler is about to write without proof of
ownership, and repair it via Cloudflare only when the ownership marker
matches. Safe to run on a schedule. Stays in dry run until DRY_RUN=false.

Guide: https://www.allanninal.dev/dns/unowned-record-overwritten/
"""
import os
import logging

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("check_record_ownership")

DNS_DOMAIN = os.environ.get("DNS_DOMAIN", "example.com")
CLOUDFLARE_API_TOKEN = os.environ.get("CLOUDFLARE_API_TOKEN", "")
CLOUDFLARE_ZONE_ID = os.environ.get("CLOUDFLARE_ZONE_ID", "")
DNS_OWNER_ID = os.environ.get("DNS_OWNER_ID", "team-a")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"


def decide_action(intended: dict, live: dict | None, owner_id: str) -> str:
    """
    intended: {"name": str, "type": str, "content": str, "owner": str}
    live: {"name": str, "type": str, "content": str, "comment": str|None} or None if absent
    owner_id: this reconciler's own owner tag, e.g. "team-a"
    Returns one of: "create", "update", "skip_conflict", "noop"
    Pure decision logic, no I/O: given the intended record, the current live record
    (or None), and this instance's owner id, decide whether it is safe to write.
    """
    if live is None:
        return "create"

    live_comment = live.get("comment") or ""
    live_owner = live_comment.split("managed-by:")[-1].strip() if "managed-by:" in live_comment else None

    if live_owner is None:
        return "skip_conflict"
    if live_owner != owner_id:
        return "skip_conflict"
    if live.get("content") == intended.get("content"):
        return "noop"
    return "update"


def run():
    # Imported lazily so the pure function above can be tested with no
    # network libraries installed at all.
    import requests

    zone_id = CLOUDFLARE_ZONE_ID
    api_token = CLOUDFLARE_API_TOKEN
    owner_id = DNS_OWNER_ID
    dry_run = DRY_RUN
    domain = DNS_DOMAIN

    # In a real run, "desired" would come from Terraform state or a
    # Kubernetes Ingress spec. Here it is one example record derived from
    # DNS_DOMAIN, so the script has something concrete to reconcile.
    desired = [
        {"name": f"app.{domain}", "type": "A", "content": "203.0.113.10", "owner": owner_id},
    ]

    headers = {"Authorization": f"Bearer {api_token}", "Content-Type": "application/json"}
    base = f"https://api.cloudflare.com/client/v4/zones/{zone_id}/dns_records"

    for intended in desired:
        resp = requests.get(base, headers=headers, params={"name": intended["name"], "type": intended["type"]}, timeout=30)
        resp.raise_for_status()
        results = resp.json().get("result", [])
        live = results[0] if results else None

        action = decide_action(intended, live, owner_id)
        log.info("Record %s (%s): action=%s", intended["name"], intended["type"], action)

        if action == "noop":
            log.info("Already correct and owned. Nothing to do.")
            continue

        if action == "skip_conflict":
            log.warning(
                "Skipping %s: live record has no matching owner marker for '%s'. "
                "Refusing to overwrite a record this reconciler does not own.",
                intended["name"], owner_id,
            )
            continue

        comment = f"managed-by:{owner_id}"
        if action == "create":
            log.info("%s create %s -> %s", "Would" if dry_run else "Will", intended["name"], intended["content"])
            if not dry_run:
                requests.post(
                    base, headers=headers,
                    json={"type": intended["type"], "name": intended["name"], "content": intended["content"], "ttl": 300, "comment": comment},
                    timeout=30,
                ).raise_for_status()
        elif action == "update":
            log.info("%s update %s -> %s", "Would" if dry_run else "Will", intended["name"], intended["content"])
            if not dry_run:
                requests.patch(
                    f"{base}/{live['id']}", headers=headers,
                    json={"content": intended["content"], "comment": comment},
                    timeout=30,
                ).raise_for_status()

    log.info("Done.")


if __name__ == "__main__":
    run()
